### PR TITLE
Add `Builder::try_parse()` method

### LINKED ISSUE #323: Add a fitler parsing function returning `Result` instead of writing errors to `stderr`
I suggest adding a new method `env_filter::filter::try_parse(&mut self, filters: &str) -> Result<&mut Self, ParsingError>`. This would imply adding a new `ParsingError` type and a small refactoring, but would make the library more convenient in some situations.

In my use case, I don't have access to the contents of `stderr` when the application is starting, and I actually don't want the application to start if the logger is configured incorrectly. Such a method would let me do this without copying the parser implementation to the application code.

If the maintainers agree that this can be added to the `env_logger`, I am happy to create a PR.

### PR DESCRIPTION
**Pull request by Maximkaaa**
Current implementation of the `Builder::parse()` method prints out specification errors to `stderr` and then just ignores them. This is fine for most console applications, but in some cases better control over what is happening is needed:

* Sometimes there is no access to `stderr`, in that case there is no way to find out what went wrong
* For some applications it's more desirable to fail on startup than to run with (partially) invalid configuration.

For such cases this PR introduces a new method `try_parse` that does the same thing as the `parse` method, but returns an `Err` in case an error in the specification is found.

This change was suggested in and fixes #323.

### CODE REVIEW

File: crates/env_filter/src/parser.rs [hunk 1]
```diff
@@ -42,20 +77,26 @@ pub(crate) fn parse_spec(spec: &str) -> (Vec<Directive>, Option<FilterOp>) {
...
+                                    part1
+                                ),
+                            )?;
                             continue;
                         }
                     }
                     _ => {
-                        eprintln!(
-                            "warning: invalid logging spec '{}', \
+                        err_or_print(
```
Comment (epage) [inline crates/env_filter/src/parser.rs:92]: Not too thrilled with `err_or_print`

I'm hoping we can instead have the following commits
- refactor commit to change `parse_spec` to return an error with `Builder::parse` taking the error and printing it as a warning
- feature commit that pulls the code out of `Builder::parse` into `Builder::try_parse`, making `Builder::parse` a wrapper around it that takes the error and prints it as a warning

File: crates/env_filter/src/parser.rs [hunk 2]
```diff
@@ -1,25 +1,60 @@
 use log::LevelFilter;
+use std::error::Error;
+use std::fmt::{Debug, Display, Formatter};
 
 use crate::Directive;
 use crate::FilterOp;
 
+/// Error during logger directive parsing process.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct ParsingError {
```
Comment (epage) [inline crates/env_filter/src/parser.rs:10]: nit: Feel like this should be named `ParseError` rather than `ParsingError`

File: crates/env_filter/src/parser.rs [hunk 3]
```diff
@@ -1,25 +1,60 @@
...
+    /// Details of what exactly was wrong in the directive.
+    pub details: String,
```
Comment (epage) [inline crates/env_filter/src/parser.rs:12]: The fields should not be public

File: crates/env_filter/src/parser.rs [hunk 4]
```diff
@@ -66,22 +107,39 @@ pub(crate) fn parse_spec(spec: &str) -> (Vec<Directive>, Option<FilterOp>) {
...
+        .map(|filter| match FilterOp::new(filter) {
+            Ok(re) => Ok(Some(re)),
+            Err(e) => err_or_print(
+                skip_errors,
+                format!("warning: invalid regex filter - {}", e),
+            )
+            .map(|_| None),
+        })
+        .transpose()?
+        .flatten();
```
Comment (epage) [inline crates/env_filter/src/parser.rs:120]: This is getting to be too complex of combinators.  I'm assuming there is a more straightforward imperative way of writing this.  Maybe refactoring the original code could also be its own commit, paving the way for the added complexity of this

File: crates/env_filter/src/parser.rs [hunk 5]
```diff
@@ -108,6 +166,11 @@ mod tests {
         assert!(filter.is_none());
     }
 
+    #[test]
+    fn try_parse_spec_invalid_crate() {
+        assert!(try_parse_spec("crate1::mod1=warn=info,crate2=debug", false).is_err());
+    }
```
Comment (epage) [inline crates/env_filter/src/parser.rs:172]: Let's add `snapbox` as a dev dependency and then you can do
```rust
let input = "crate1::mod1=warn=info,crate2=debug";
let actual = try_parse_spec(input).unwrap_err();
assert_data_eq!(actual, str![]);
```
and then use snapshot testing to fill in the expected result.

This will ensure we test the quality of the messages for a given string *and* can easily update the tests if we work to improve the quality of the messages

File: crates/env_filter/src/parser.rs (hunk 1 above)
Comment (Maximkaaa) [reply crates/env_filter/src/parser.rs]: I agree, this was not the best solution. But `parse_spec` cannot just return `Result`, as `Builder::parse` method needs the valid part of the spec even in case of an error. Also, it is possible to have multiple errors in the same spec, so just bailing out with the first error would also not work.

Instead, I've added a new `ParseResult` struct, that contains original returned values plus the encountered errors, and made the caller decide, what they want to do with these errors. I believe this approach looks quite clean now.

File: crates/env_filter/src/parser.rs (hunk 2 above)
Comment (Maximkaaa) [reply crates/env_filter/src/parser.rs]: Done

File: crates/env_filter/src/parser.rs (hunk 3 above)
Comment (Maximkaaa) [reply crates/env_filter/src/parser.rs]: Done

File: crates/env_filter/src/parser.rs (hunk 4 above)
Comment (Maximkaaa) [reply crates/env_filter/src/parser.rs]: With the new `ParseResult` this code seems much cleaner.

File: crates/env_filter/src/parser.rs (hunk 5 above)
Comment (Maximkaaa) [reply crates/env_filter/src/parser.rs]: Thanks for the pointers to the `snapbox`. It's really convenient!

File: crates/env_filter/Cargo.toml [hunk 6]
```diff
@@ -18,11 +18,11 @@ rustdoc-args = ["--cfg", "docsrs"]
 
 [package.metadata.release]
 pre-release-replacements = [
-  {file="CHANGELOG.md", search="Unreleased", replace="{{version}}", min=1},
```
Comment (Maximkaaa) [inline crates/env_filter/Cargo.toml:21]: Sorry, missed the `toml` file being autoformatted... Shall I try to revert this?

File: crates/env_filter/src/parser.rs [hunk 7]
```diff
@@ -3,23 +3,42 @@ use log::LevelFilter;
...
+pub(crate) fn parse_spec(spec: &str) -> ParseResult {
+    let mut result = ParseResult::default();
 
     let mut parts = spec.split('/');
     let mods = parts.next();
     let filter = parts.next();
     if parts.next().is_some() {
-        eprintln!(
+        result.add_error(format!(
             "warning: invalid logging spec '{}', \
```
Comment (epage) [inline crates/env_filter/src/parser.rs]: The `warning:` prefix should be in the caller

File: crates/env_filter/Cargo.toml (hunk 6 above)
Comment (epage) [reply crates/env_filter/Cargo.toml:21]: Please do so as its unrelated to tis commit

File: crates/env_filter/src/parser.rs (hunk 7 above)
Comment (epage) [reply crates/env_filter/src/parser.rs]: Ah, you did that in the follow up but I  feel like that should belong in this commit

File: crates/env_filter/src/parser.rs [hunk 8]
```diff
@@ -22,8 +24,36 @@ impl ParseResult {
...
+            Err(ParseError {
+                details: self.errors,
+            })
+        }
+    }
 }
 
+/// Error during logger directive parsing process.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct ParseError {
```
Comment (epage) [inline crates/env_filter/src/parser.rs:41]: An aliases commit makes reference to `ParseError`

File: crates/env_filter/src/parser.rs [hunk 9]
```diff
@@ -22,8 +24,36 @@ impl ParseResult {
     fn add_error(&mut self, message: String) {
         self.errors.push(message);
     }
+
+    pub(crate) fn ok(self) -> Result<(Vec<Directive>, Option<FilterOp>), ParseError> {
+        if self.errors.is_empty() {
+            Ok((self.directives, self.filter))
+        } else {
+            Err(ParseError {
+                details: self.errors,
```
Comment (epage) [inline crates/env_filter/src/parser.rs:33]: Unsure how I feel about printing the details comma separated.  Should we just return the first?

**Comment by epage** [comment]
Thanks for moving this forward!

File: crates/env_filter/src/parser.rs (hunk 9 above)
Comment (Maximkaaa) [reply crates/env_filter/src/parser.rs:33]: I would argue that the more details an error message provides, the better. In this case, we already allocate for `String` message in the error, so giving more details is doesn't have downsides in my opinion. It's often frustrating when a system returns only first error of many when it easily can save you several iterations of fixing your code by simply telling you all your errors right away.

If you find `Vec` in the error struct inappropriate, we could join strings here instead of doing it in the `Display` (although I think it's better to keep messages separated in the error struct).

File: crates/env_filter/Cargo.toml (hunk 6 above)
Comment (Maximkaaa) [reply crates/env_filter/Cargo.toml]: Done

File: crates/env_filter/src/parser.rs (hunk 7 above)
Comment (Maximkaaa) [reply crates/env_filter/src/parser.rs]: Moved it into the first commit

File: crates/env_filter/src/parser.rs (hunk 8 above)
Comment (Maximkaaa) [reply crates/env_filter/src/parser.rs:41]: Fixed

File: crates/env_filter/src/parser.rs [hunk 10]
```diff
@@ -98,17 +86,26 @@ pub(crate) fn parse_spec(spec: &str) -> ParseResult {
...
 mod tests {
+    use crate::ParseError;
     use log::LevelFilter;
+    use snapbox::{assert_data_eq, str, Data, IntoData};
 
     use super::{parse_spec, ParseResult};
 
+    impl IntoData for ParseError {
+        fn into_data(self) -> Data {
+            format!("{self}").into_data()
```
[the remark this answers is not in the document]
Comment (Maximkaaa) [reply crates/env_filter/src/parser.rs]: Done

**Comment by Maximkaaa** [comment]
I fixed all the comments except the last one, since I think that the current approach is better. If you insist on returning just one error string, I will edit it also.

File: crates/env_filter/CHANGELOG.md [hunk 11]
```diff
@@ -1,14 +1,22 @@
 # Change Log
+
 All notable changes to this project will be documented in this file.
 
 The format is based on [Keep a Changelog](http://keepachangelog.com/)
 and this project adheres to [Semantic Versioning](http://semver.org/).
 
 <!-- next-header -->
+
 ## [Unreleased] - ReleaseDate
```
Comment (epage) [inline crates/env_filter/CHANGELOG.md:11]: Please revert these unrelated changes

File: crates/env_filter/src/parser.rs (hunk 9 above)
Comment (epage) [reply crates/env_filter/src/parser.rs:33]: I'm not worried about performance but the usability of concatenating the errors. imo we should only return one.  If someone finds a concrete use case where this is inhibiting usability enough, let's evaluate a "error recovery" API that returns a `Vec<ParseError>` (or some other related approach).

File: crates/env_filter/CHANGELOG.md (hunk 11 above)
Comment (Maximkaaa) [reply crates/env_filter/CHANGELOG.md]: Sorry for that. Done.

File: crates/env_filter/src/parser.rs (hunk 9 above)
Comment (Maximkaaa) [reply crates/env_filter/src/parser.rs]: All right, I've made the error contain the first error message.

**Comment by epage** [comment]
Looks like the final commit needs `cargo fmt` run on it

**Comment by Maximkaaa** [comment]
> Looks like the final commit needs `cargo fmt` run on it

Done.

**Comment by epage** [comment]
Thanks!  This is now released.
